### PR TITLE
[CB] add min batch size of 2 in decode

### PR DESCRIPTION
**Pull request by nikolaospapandreou**
This PR adds support for minimum batch size of 2 in decode steps.

### CODE REVIEW

File: vllm_spyre/v1/worker/spyre_model_runner.py [hunk 1]
```diff
@@ -612,13 +613,26 @@ def _update_states(self, scheduler_output):
...
         self,
         new_requests: list[NewRequestData],
     ) -> ModelForwardInputs:
         assert len(new_requests) > 0
         input_token_list: list[torch.Tensor] = []
 
+        # free the blocks used for padding to minimum decode batch size of 2
+        if self.dummy_req_ids2blocks:
+            for freed_block in self.dummy_req_ids2blocks:
+                self.free_blocks.appendleft(freed_block)
```
Comment (sducouedic) [inline vllm_spyre/v1/worker/spyre_model_runner.py:633]: I think `appendleft` won't do exactly what you want. It will append the values from `dummy_req_ids2blocks` to the left, but in reverse order. e.g. `[6, 7] --> appendleft --> [7, 6, 8, 9, 10, ...]`. I would suggest to use `append` simply, just as you did in line 620. Or iterate the `dummy_req_ids2blocks` in reverse order (and do the same in 620)

**Review by sducouedic — APPROVED** [review]
LGTM (just a minor comment)
nice changes 👍

File: vllm_spyre/v1/worker/spyre_model_runner.py [hunk 2]
```diff
@@ -755,6 +769,27 @@ def _prepare_decode(
...
+            dummy_req_indices = torch.zeros(1, dtype=torch.bool, device="cpu")
+            self.model.indices = torch.cat(
+                (self.model.indices, dummy_req_indices), -1)
+            assert self.model.indices.size(dim=0) == 2
+
+            n = self.tkv + 1
+            d = self.BLOCK_SIZE
+            num_blocks = (n + d - 1) // d
+            for i in range(num_blocks - len(self.dummy_req_ids2blocks)):
+                self.dummy_req_ids2blocks.append(self.free_blocks.popleft())
```
Comment (joerunde) [inline vllm_spyre/v1/worker/spyre_model_runner.py:783]: Is this popping free blocks for every decode iteration of a batch with a single request? It looks like `len(input_tokens) == 1` will always be true when there's only a single request running, since `cached_requests` comes from the scheduler and won't have our dummy request in it

Comment (joerunde) [reply vllm_spyre/v1/worker/spyre_model_runner.py:783]: ahhh I see the `num_blocks - len(self.dummy_req_ids2blocks)` is only popping ones as the tkv increases 👍

File: vllm_spyre/v1/worker/spyre_model_runner.py [hunk 3]
```diff
@@ -612,13 +613,26 @@ def _update_states(self, scheduler_output):
...
     ) -> ModelForwardInputs:
         assert len(new_requests) > 0
         input_token_list: list[torch.Tensor] = []
 
+        # free the blocks used for padding to minimum decode batch size of 2
+        if self.dummy_req_ids2blocks:
+            for freed_block in self.dummy_req_ids2blocks:
+                self.free_blocks.appendleft(freed_block)
+            self.dummy_req_ids2blocks = []
+
```
[the remark this answers is not in the document]
Comment (sducouedic) [reply vllm_spyre/v1/worker/spyre_model_runner.py:635]: I guess this would do the cleanup when the last requests finishes?

**Comment by yannicks1** [comment]
> I am not completely sure if this won't automatically be handled correctly in the subsequent _prepare_decode, but probably in reduce_left_padding we should also free the dummy blocks

good point! this will be solved when doing the removal of the left padding in every step (decode **and** prefill). I am taking care of this on this branch: https://github.com/vllm-project/vllm-spyre/tree/ysc-refactor-left-padding-stripping
